consensus/istanbul/core/vrank.go

```diff
@@ -0,0 +1,240 @@
+	vrankAvgCommitArrivalTimeWithinQuorumGauge = metrics.NewRegisteredGauge("vrank/avg_commit_within_quorum", nil)
+	vrankLastCommitArrivalTimeGauge            = metrics.NewRegisteredGauge("vrank/last_commit", nil)
+
+	vrankDefaultThreshold = "300ms"
+
+	vrank *Vrank
+)
+
+const (
+	vrankArrivedEarly = iota
+	vrankArrivedLate
+	vrankNotArrived
+)
+
+const (
+	vrankNotArrivedPlaceholder = -1
+)
+
+func NewVrank(view istanbul.View, committee istanbul.Validators) *Vrank {
+	threshold, _ := time.ParseDuration(vrankDefaultThreshold)
+	return &Vrank{
+		startTime:            time.Now(),
+		view:                 view,
+		committee:            committee,
+		threshold:            threshold,
+		commitArrivalTimeMap: make(map[common.Address]time.Duration),
+	}
+}
+
+func (v *Vrank) TimeSinceStart() time.Duration {
+	return time.Now().Sub(v.startTime)
+}
+
+func (v *Vrank) AddCommit(msg *istanbul.Subject, src istanbul.Validator) time.Duration {
+	if v.isTargetCommit(msg, src) {
+		t := v.TimeSinceStart()
+		v.commitArrivalTimeMap[src.Address()] = t
+		return t
+	}
+	return -1
+}
+
+func (v *Vrank) HandleCommitted(blockNum *big.Int) {
+	if v.view.Sequence.Cmp(blockNum) != 0 {
+		return
+	}
+
+	committedTime := v.TimeSinceStart()
+	if v.threshold > committedTime {
+		v.threshold = committedTime
+	}
+
+	vrankQuorumCommitArrivalTimeGauge.Update(int64(committedTime))
+	if len(v.commitArrivalTimeMap) != 0 {
+		sum := int64(0)
+		for _, v := range v.commitArrivalTimeMap {
+			sum += int64(v)
+		}
+		avg := sum / int64(len(v.commitArrivalTimeMap))
+		vrankAvgCommitArrivalTimeWithinQuorumGauge.Update(avg)
+	}
+}
+
+// Log logs accumulated data in a compressed form
+func (v *Vrank) Log() (string, []string) {
+	var (
+		serialized = serialize(v.committee, v.commitArrivalTimeMap)
+		assessed   = assessBatch(serialized, v.threshold)
+		compressed = compress(assessed)
+		bitmap     = hex.EncodeToString(compressed)
+
+		lateCommits = make([]time.Duration, 0)
+		lastCommit  = time.Duration(0)
+	)
+
+	for _, t := range serialized {
+		if assess(t, v.threshold) == vrankArrivedLate {
+			lateCommits = append(lateCommits, t)
+		}
+	}
+
+	// lastCommit = max(lateCommits)
+	for _, t := range lateCommits {
+		if lastCommit < t {
+			lastCommit = t
+		}
+	}
+	if lastCommit != time.Duration(0) {
+		vrankLastCommitArrivalTimeGauge.Update(int64(lastCommit))
+	}
+
+	lateCommitsStrArr := encodeDurationBatch(lateCommits)
+
+	logger.Info("VRank", "seq", v.view.Sequence.Int64(),
+		"round", v.view.Round.Int64(),
+		"bitmap", bitmap,
+		"late", lateCommitsStrArr,
+	)
+	return bitmap, lateCommitsStrArr
+}
+
+func (v *Vrank) isTargetCommit(msg *istanbul.Subject, src istanbul.Validator) bool {
+	if msg.View.Cmp(&v.view) != 0 {
+		return false
+	}
+	_, ok := v.commitArrivalTimeMap[src.Address()]
+	if ok {
+		return false
+	}
+	return true
+}
+
+// assess determines if given time is early, late, or not arrived
+func assess(t, threshold time.Duration) int {
```

Comment:
How about changing the type `int` to `uint8`? Then, `vrankNotArrivedPlaceholder` should be changed accordingly.

Reply:
fixed in c34ac7bb. there's no need to change `vrankNotArrivedPlaceholder` though

Reply:
I'm sorry. I omitted to mention that defines a new type for it (e.g., type Bitmap []uint8). You may consider it later.